plugins/timertop/timerinfo.cpp

```diff
@@ -30,194 +30,253 @@
 {
-    if (m_type == QObjectType)
-        return QStringLiteral("N/A");
-
-    int max = 0;
-    for (int i = 0; i < m_timeoutEvents.size(); i++) {
-        const TimeoutEvent &event = m_timeoutEvents.at(i);
-        if (event.executionTime > max)
-            max = event.executionTime;
-    }
-    return QString::number(max);
+    return m_timerAddress == reinterpret_cast<quint64>(timer);
 }
 
-int TimerInfo::totalWakeups() const
+bool TimerId::operator==(int timerId) const
 {
-    return m_totalWakeups;
+    return m_timerId == timerId;
 }
 
-QString TimerInfo::state() const
+bool TimerId::operator<(const TimerId &other) const
 {
-    switch (type()) {
-    case QTimerType:
-    {
-        const QTimer *t = timer();
-        if (!t)
-            return TimerModel::tr("None");
-        if (!t->isActive())
-            return TimerModel::tr("Inactive");
-        if (t->isSingleShot())
-            return TimerModel::tr("Singleshot (%1 ms)").arg(t->interval());
-        return TimerModel::tr("Repeating (%1 ms)").arg(t->interval());
-    }
-    case QQmlTimerType:
-    {
-        const QObject *obj = timerObject();
-        if (!obj)
-            return TimerModel::tr("None");
-        const int interval = obj->property("interval").toInt();
-        if (!obj->property("running").toBool())
-            return TimerModel::tr("Inactive (%1 ms)").arg(interval);
-        if (obj->property("repeat").toBool())
-            return TimerModel::tr("Repeating (%1 ms)").arg(interval);
-        return TimerModel::tr("Singleshot (%1 ms)").arg(interval);
-    }
-    case QObjectType:
-        return QStringLiteral("N/A");
+    if (m_timerAddress != 0) {
+        if (other.m_timerAddress != 0)
+            return m_timerAddress < other.m_timerAddress;
+        return other.m_timerId != -1 ? m_timerAddress < quint64(other.m_timerId) : false;
+    } else if (m_timerId != -1) {
+        if (other.m_timerId != -1)
+            return m_timerId < other.m_timerId;
+        return other.m_timerAddress != 0 ? quint64(m_timerId) < other.m_timerAddress : false;
     }
 
     Q_ASSERT(false);
-    return QString();
+    return false;
 }
 
-void TimerInfo::removeOldEvents()
+TimerIdInfo::TimerIdInfo(const TimerId &id, QObject *receiver)
+    : timerId(-1)
+    , totalWakeups(0)
+    , lastReceiverAddress(0)
 {
-    if (m_timeoutEvents.size() > maxTimeoutEvents)
-        m_timeoutEvents.removeFirst();
+    update(id, receiver);
 }
 
-void TimerInfo::setLastReceiver(QObject *receiver)
+void TimerIdInfo::update(const TimerId &id, QObject *receiver)
 {
-    m_lastReceiver = receiver;
-}
+    Q_ASSERT(id.isValid());
+    QObject *object = receiver ? receiver : id.timerObject();
+    // We can't lock TimerModel::m_mutex and Probe::objectLock() on the same thread at the same time
+    QSharedPointer<QMutexLocker> locker;
+    if (object->thread() != QThread::currentThread())
+        locker = QSharedPointer<QMutexLocker>(new QMutexLocker(Probe::objectLock()));
```

Comment:
Even implementation details aside, locking a mutex only in some threads would seem problematic from a correctness point-of-view. The comment just reads like "otherwise it deadlocks" to me, if there are pre-conditions that make this code correct (like objectLock already being held by the caller in the main thread), this should be mentioned.